File: CPP/src/engine/math/Matrix4f.cpp
//
// Created by Mnenmenth on 1/8/2016.
//

#define PI 3.14159265358979323846

#include <math.h>
#include "Matrix4f.h"

Matrix4f::Matrix4f() {

}

Matrix4f Matrix4f::initIdentity(){
    for(int i = 0; i <= 3; i++){
        for(int j = 0; j <= 3; j++){
            matrix[i][j] = 0;
        }
    }

    for(int i = 0; i <= 3; i++){
        matrix[i][i] = 1;
    }

    return *this;
}


Matrix4f Matrix4f::initTranslation(float x, float y, float z) {
    for(int i = 0; i <= 3; i++){
        for(int j = 0; j <= 3; j++){
            matrix[i][j] = 0;
        }
    }

    for(int i = 0; i <= 3; i++){
        matrix[i][i] = 1;
    }
    matrix[0][3] = x;
    matrix[1][3] = y;
    matrix[2][3] = z;
    return *this;
}

Matrix4f Matrix4f::initRotation(float x, float y, float z) {
    Matrix4f rx = Matrix4f();
    Matrix4f ry = Matrix4f();
    Matrix4f rz = Matrix4f();

    float xRad = (float)(x * PI / 2);
    float yRad = (float)(y * PI / 2);
    float zRad = (float)(z * PI / 2);

    for(int i = 0; i <= 3; i++){
        rx.matrix[i][i] = 1;
        ry.matrix[i][i] = 1;
        rz.matrix[i][i] = 1;
    }
    
    rx.matrix[1][1] = (float)cos(xRad);
    rx.matrix[1][1] = -(float)cos(xRad);
    rx.matrix[1][1] = (float)cos(xRad);
    rx.matrix[1][1] = (float)cos(xRad);

    ry.matrix[1][1] = (float)cos(yRad);
    ry.matrix[1][1] = (float)sin(yRad);
    ry.matrix[1][1] = -(float)sin(yRad);
    ry.matrix[1][1] = (float)cos(yRad);

    rz.matrix[1][1] = (float)cos(zRad);
    rz.matrix[1][1] = -(float)sin(zRad);
    rz.matrix[1][1] = (float)sin(zRad);
    rz.matrix[1][1] = (float)cos(zRad);

    matrix = rz.mult(ry.mult(rx)).matrix;
    return *this;
}

Matrix4f Matrix4f::initScale(float x, float y, float z) {
    for(int i = 0; i <= 3; i++){
        for(int j = 0; j <= 3; j++){
            matrix[i][j] = 0;
        }
    }
    matrix[0][0] = x;
    matrix[1][1] = y;
    matrix[2][2] = z;
    matrix[3][3] = 1;
    return *this;
}


Matrix4f Matrix4f::mult(Matrix4f m) {
    Matrix4f res = Matrix4f();
    for(int i = 0; i < 4; i++){
        for(int j = 0; j < 4; j++){
            res.set(i, j, matrix[i][0] * m.get(0, j) +
                          matrix[i][1] * m.get(1, j) +
                          matrix[i][2] * m.get(2, j) +
                          matrix[i][3] * m.get(3, j));
        }
    }
    return res;
}

std::vector<std::vector<float>> Matrix4f::getM() {
    return matrix;
}

void Matrix4f::setM(std::vector<std::vector<float>> matrix) {
    matrix = matrix;
}

float Matrix4f::get(int x, int y) {
    return matrix[x][y];
}

void Matrix4f::set(int x, int y, float value) {
    matrix[x][y] = value;
}

File: JAVA/src/main/scala/engine/math/Matrix4f.scala
package engine.math

import scala.collection.mutable.ListBuffer

/**
  * Created by Mnenmenth Alkaborin
  * Please refer to LICENSE file if included
  * for licensing information
  * https://github.com/Mnenmenth
  */
class Matrix4f {

  var matrix = Array.ofDim[Float](4,4)

  def initIdentity(): Matrix4f = {
    /*matrix(0)(0) = 1; matrix(0)(1) = 0; matrix(0)(2) = 0; matrix(0)(3) = 0
    matrix(1)(0) = 0; matrix(1)(1) = 1; matrix(1)(2) = 0; matrix(1)(3) = 0
    matrix(2)(0) = 0; matrix(2)(1) = 0; matrix(2)(2) = 1; matrix(2)(3) = 0
    matrix(3)(0) = 0; matrix(3)(1) = 0; matrix(3)(2) = 0; matrix(3)(3) = 1
    */
    for(i <- 0 to 3) for(j <- 0 to 3) matrix(i)(j) = 0
    for(i <- 0 to 3) matrix(i)(i) = 1
    this
  }

  def initTranslation(x: Float, y: Float, z: Float): Matrix4f = {
    /*matrix(0)(0) = 1; matrix(0)(1) = 0; matrix(0)(2) = 0; matrix(0)(3) = x
    matrix(1)(0) = 0; matrix(1)(1) = 1; matrix(1)(2) = 0; matrix(1)(3) = y
    matrix(2)(0) = 0; matrix(2)(1) = 0; matrix(2)(2) = 1; matrix(2)(3) = z
    matrix(3)(0) = 0; matrix(3)(1) = 0; matrix(3)(2) = 0; matrix(3)(3) = 1
    */
    for(i <- 0 to 3) for(j <- 0 to 3) matrix(i)(j) = 0
    for(i <- 0 to 3) matrix(i)(i) = 1
    matrix(0)(3) = x
    matrix(1)(3) = y
    matrix(2)(3) = z

    this
  }

  def initRotation(x: Float, y: Float, z: Float): Matrix4f = {

    val rx = new Matrix4f
    val ry = new Matrix4f
    val rz = new Matrix4f

    val xRad = Math.toRadians(x).toFloat
    val yRad = Math.toRadians(y).toFloat
    val zRad = Math.toRadians(z).toFloat

    for(i <- 0 to 3){
      rx.matrix(i)(i) = 1
      ry.matrix(i)(i) = 1
      rz.matrix(i)(i) = 1
    }

    rx.matrix(1)(1) = Math.cos(xRad).toFloat
    rx.matrix(1)(2) = -Math.cos(xRad).toFloat
    rx.matrix(2)(1) = Math.cos(xRad).toFloat
    rx.matrix(2)(2) = Math.cos(xRad).toFloat

    ry.matrix(0)(0) = Math.cos(yRad).toFloat
    ry.matrix(0)(2) = Math.sin(yRad).toFloat
    ry.matrix(2)(0) = -Math.sin(yRad).toFloat
    ry.matrix(2)(2) = Math.cos(yRad).toFloat

    rz.matrix(0)(0) = Math.cos(zRad).toFloat
    rz.matrix(0)(1) = -Math.sin(zRad).toFloat
    rz.matrix(1)(0) = Math.sin(zRad).toFloat
    rz.matrix(1)(1) = Math.cos(zRad).toFloat

    matrix = rz.mult(ry.mult(rx)).matrix
    this
  }

  def initScale(x: Float, y: Float, z: Float): Matrix4f = {
    for(i <- 0 to 3) for(j <- 0 to 3) matrix(i)(j) = 0
    matrix(0)(0) = x
    matrix(1)(1) = y
    matrix(2)(2) = z
    matrix(3)(3) = 1
    this
  }

  def mult(m: Matrix4f): Matrix4f = {
    val res = new Matrix4f
    for(i <- 0 until 4){
      for(j <- 0 until 4){
        res.set(i, j, matrix(i)(0) * m.get(0, j) +
                      matrix(i)(1) * m.get(1, j) +
                      matrix(i)(2) * m.get(2, j) +
                      matrix(i)(3) * m.get(3, j))
      }
    }
    res
  }

  def get(x: Int, y: Int): Float = matrix(x)(y)
  def set(x: Int, y: Int, value: Float): Unit = matrix(x)(y) = value

}
